### ISSUE TITLE
2d pretty printer ignores successive empty lines

### ISSUE THREAD

**Issue by rtoy** [rtoy/maxima#2708]
Imported from SourceForge on 2024-07-06 18:13:28
Created by **[robert_dodier](https://sourceforge.net/u/robert_dodier/)** on 2024-04-11 23:09:00
Original: https://sourceforge.net/p/maxima/bugs/4287

---

Lines which are empty (i.e., comprising only successive newline characters and no others) are parsed correctly, and displayed correctly by the 1d printer, but the 2d printer doesn't display the empty lines.
```
(%i1) display2d: false $
(%i2) "aaaaa



bbbbb";
(%o2) "aaaaa



bbbbb"
(%i3) display2d: true $
(%i4) "aaaaa



bbbbb";
(%o4) aaaaa
bbbbb
```
The string %o4 contains the empty lines as expected (character code 10):
```
(%i5) :lisp (mapcar #'char-code (coerce $%o4 'list))
(97 97 97 97 97 10 10 10 10 98 98 98 98 98)
```
Working with SBCL 2.3.7 + commit ff883f3. 

Looks like DIMENSION-STRING in the display code code puts the empty lines into its return value. 
```
(%i5) :lisp (trace dimension-string)
(DIMENSION-STRING)
(%i5) %o4;
  0: (MAXIMA::DIMENSION-STRING (#\% #\o #\5) (#\())
  0: DIMENSION-STRING returned (#\5 #\o #\% #\()
  0: (MAXIMA::DIMENSION-STRING (#\a #\a #\a #\a #\a #\Newline #\Newline #\Newline #\Newline #\b #\b #\b #\b #\b) ((0 0) #\  #\) #\5 #\o #\% #\())
(%o5) aaaaa
  0: DIMENSION-STRING returned (#\b #\b #\b #\b #\b)
bbbbb
```
But DRAW-LINEAR seems to ignore the empty lines.
```
(%i6) :lisp (untrace) (trace draw-linear)

(DRAW-LINEAR)
(%i6) %o4;
  0: (MAXIMA::DRAW-LINEAR (#\( #\% #\o #\6 #\) #\  (0 0) #\a #\a #\a #\a #\a) 0 0)
    1: (MAXIMA::DRAW-LINEAR NIL 0 6)
    1: DRAW-LINEAR returned 6
  0: DRAW-LINEAR returned 11
(%o6) aaaaa
  0: (MAXIMA::DRAW-LINEAR NIL 0 0)
  0: DRAW-LINEAR returned 0
  0: (MAXIMA::DRAW-LINEAR NIL 0 0)
  0: DRAW-LINEAR returned 0
  0: (MAXIMA::DRAW-LINEAR NIL 0 0)
  0: DRAW-LINEAR returned 0
  0: (MAXIMA::DRAW-LINEAR (#\b #\b #\b #\b #\b) 0 0)
  0: DRAW-LINEAR returned 5
bbbbb
```
But if there is at least one space character on the line, DRAW-LINEAR displays something.
```
(%i7) "aaaa
 
 
 
bbb";
  0: (MAXIMA::DRAW-LINEAR (#\( #\% #\o #\7 #\) #\  (0 0) #\a #\a #\a #\a) 0 0)
    1: (MAXIMA::DRAW-LINEAR NIL 0 6)
    1: DRAW-LINEAR returned 6
  0: DRAW-LINEAR returned 10
(%o7) aaaa
  0: (MAXIMA::DRAW-LINEAR (#\ ) 0 0)
  0: DRAW-LINEAR returned 1
 
  0: (MAXIMA::DRAW-LINEAR (#\ ) 0 0)
  0: DRAW-LINEAR returned 1
 
  0: (MAXIMA::DRAW-LINEAR (#\ ) 0 0)
  0: DRAW-LINEAR returned 1
 
  0: (MAXIMA::DRAW-LINEAR (#\b #\b #\b) 0 0)
  0: DRAW-LINEAR returned 3
bbb
```
I don't think I want to introduce a nonexistent character to force DRAW-LINEAR to display it; it seems better to special-case the arguments NIL 0 0 and just increment the line counter or whatever. However, glancing at the code, it's not yet entirely clear exact what to do.

**Comment by rtoy** [comment]
Imported from SourceForge on 2024-07-06 18:13:29
Created by **[robert_dodier](https://sourceforge.net/u/robert_dodier/)** on 2024-04-11 23:10:46
Original: https://sourceforge.net/p/maxima/bugs/4287/#8aad

---

It occurs to me that we would have to watch out if `(draw-linear nil 0 0)` is ever called in any situation other than to display an empty line.

**Comment by rtoy** [comment]
Imported from SourceForge on 2024-07-06 18:13:33
Created by **[robert_dodier](https://sourceforge.net/u/robert_dodier/)** on 2024-04-14 03:19:32
Original: https://sourceforge.net/p/maxima/bugs/4287/#28d2

---

- **status**: open --> closed

**Comment by rtoy** [comment]
Imported from SourceForge on 2024-07-06 18:13:36
Created by **[robert_dodier](https://sourceforge.net/u/robert_dodier/)** on 2024-04-14 03:19:32
Original: https://sourceforge.net/p/maxima/bugs/4287/#a8ef

---

Fixed by commit [ 5a08d75 ].
